### PR TITLE
Simplified project structure

### PR DESCRIPTION
**Pull request by jroper**

### CODE REVIEW

**Comment by dwijnand** [comment]
Needs rebasing (or merging master in).

File: src/main/scala/com/typesafe/sbt/jse/SbtJsEngine.scala [hunk 1]
```diff
@@ -22,54 +19,56 @@ object JsEngineImport {
...
   import autoImport._
   import JsEngineKeys._
 
   /**
-   * Convert an engine type enum to an actor props.
-   */
-  def engineTypeToProps(engineType: EngineType.Value, command: Option[File], env: Map[String, String]) = {
+    * Convert an engine type enum to an engine.
+    */
+  def engineTypeToEngine(engineType: EngineType.Value, command: Option[File], env: Map[String, String]): Engine = {
```
Comment (dwijnand) [inline src/main/scala/com/typesafe/sbt/jse/SbtJsEngine.scala]: Does this need to be public?

File: src/main/scala/com/typesafe/sbt/jse/SbtJsTask.scala [hunk 2]
```diff
@@ -382,52 +353,52 @@ object SbtJsTask extends AutoPlugin {
...
+    * }
+    * }}}
+    *
+    * @param state       The SBT state.
+    * @param engineType  The type of engine to use.
+    * @param command     An optional path to the engine.
+    * @param nodeModules The node modules to provide (if the JavaScript engine in use supports this).
+    * @param shellSource The script to execute.
+    * @param args        The arguments to pass to the script.
+    * @param timeout     The maximum amount of time to wait for the script to finish.
```
Comment (dwijnand) [inline src/main/scala/com/typesafe/sbt/jse/SbtJsTask.scala]: no timeout

File: src/main/scala/com/typesafe/sbt/jse/SbtJsTask.scala [hunk 3]
```diff
@@ -382,52 +353,52 @@ object SbtJsTask extends AutoPlugin {
...
-   *         last thing it does, sending a DLE character (0x10) followed by some JSON to std out.
-   */
+    * Execute some arbitrary JavaScript.
+    *
+    * This method is intended to assist in building SBT tasks that execute generic JavaScript.  For example:
+    *
+    * {{{
+    * myTask := {
+    *   executeJs(state.value, engineType.value, Seq((nodeModules in Plugin).value.getCanonicalPath,
+    *     baseDirectory.value / "path" / "to" / "myscript.js", Seq("arg1", "arg2"), 30.seconds)
```
Comment (dwijnand) [inline src/main/scala/com/typesafe/sbt/jse/SbtJsTask.scala]: without the timeout

File: src/main/scala/com/typesafe/sbt/jse/SbtJsTask.scala [hunk 4]
```diff
@@ -382,52 +353,52 @@ object SbtJsTask extends AutoPlugin {
...
-        import ExecutionContext.Implicits.global
-        Await.result(
-          executeJsOnEngine(engine, shellSource, args, m => state.log.error(m), m => state.log.info(m)),
-          timeout
-        )
-    }
+    executeJsOnEngine(engine, shellSource, args, m => state.log.error(m), m => state.log.info(m))
   }
 
+  @deprecated("Use the other executeJs instead", "1.3.0")
```
Comment (dwijnand) [inline src/main/scala/com/typesafe/sbt/jse/SbtJsTask.scala]: Why is this preserved?

**Comment by mkurz** [comment]
> This depends on [sbt/sbt-web#169](https://github.com/sbt/sbt-web/pull/169) being merged.

@jroper [sbt/sbt-web#169](https://github.com/sbt/sbt-web/pull/169) got merged already.

**Comment by mkurz** [comment]
This LGTM, did quite some reviewing yesterday late night, including all the jse and npm sources. All test pass locally, I am confident this PR should be ok.

**Comment by platypii** [comment]
Hey I am maintainer of an sbt plugin (sbt-typescript) which depended on the `parallelism` option to have acceptable performance. The option was removed in this PR without much/any discussion?

Without option `parallelism := 1`, it creates a new `node` process for _every_ source file which I am compiling. This takes the runtime on a project with 151 typescript files from 8 seconds, to 17 minutes!

Is there an alternative set of options I can use to pass all source files to the js engine at once? Or some discussion about why this useful feature was removed?

**Comment by mkurz** [comment]
@platypii It was removed because lot's of functionally was based on akka, which did not make much sense in general. We might be able to make the `parallelism` setting to something useful again, would you be interested taking a look yourself and provide a pr?

File: src/main/scala/com/typesafe/sbt/jse/SbtJsTask.scala [hunk 5]
```diff
@@ -264,72 +241,61 @@ object SbtJsTask extends AutoPlugin {
...
+    val results: (Set[File], Seq[Problem]) = incremental.syncIncremental((config / streams).value.cacheDirectory / "run", sources) {
       modifiedSources: Seq[File] =>
 
-        if (modifiedSources.size > 0) {
+        if (modifiedSources.nonEmpty) {
 
           logger.info(s"$taskMsg on ${modifiedSources.size} source(s)")
 
-          val resultBatches: Seq[Future[(FileOpResultMappings, Seq[Problem])]] = {
-            val sourceBatches = (modifiedSources grouped Math.max(modifiedSources.size / taskParallelism, 1)).toSeq
```
Comment (mkurz) [inline src/main/scala/com/typesafe/sbt/jse/SbtJsTask.scala:310]: @platypii This is were to start I guess.

**Comment by platypii** [comment]
Absolutely I'm happy to take a look and potentially open a PR. Just wanted to understand the thinking first, so thank you for the response!

It's not really the parallelism that I need. Rather, I would like the ability to aggregate all the sources into one call to node, instead of many calls. Maybe there is an easier way to do this, without needing akka, or some other heavyweight mechanism. I'll take a look.

**Comment by mkurz** [comment]
Looking at the code I do think this was just a mistake when transitioning the code away from Akka. _Right now_ I think all sources should be processed at once actually...

**Comment by mkurz** [comment]
@platypii So I came with a patch myself which hopefully should fix the flaw:

- #120

I released that as version `1.3.5-M1`, can you give that a shot with your typescript files and if it works I will promote it as `1.3.5` final, handling it as a bug fix, since that what it is IMHO:
* https://github.com/sbt/sbt-js-engine/releases/tag/1.3.5-M1
* https://repo1.maven.org/maven2/com/github/sbt/sbt-js-engine_2.12_1.0/ (should be available on Maven Central soon :crossed_fingers:)

I think one of the reasons processing is so slow was of course that for each file a own process got started, but also that these processes run sequentially since it seems the code is blocking... Also the math would fit: 151 typescript files, roughly 8 seconds per file brings us to 1208 seconds total which is ~20 minutes (compared to your reported 17 minutes)
IMHO we don't need a parallelism config _at the moment_, but just process all files at once. We will see if someone desires to introduce "real" parallelism, by running multiple processes in _parallel_ (that's why I fixed the code very "simple" for now so we could introduce parallelism again later if needed).

**Comment by platypii** [comment]
@mkurz thank you for the quick turn around! I have confirmed that this does in fact fix the performance regression I was seeing, and that my typescript compile now finishes in mere seconds.

Looking forward to seeing an official release with this fix. THANK YOU! 🙌

**Comment by mkurz** [comment]
Happy it works for you, just released 1.3.5, enjoy :wink:

**Pull request merged**